### DISCUSSION TITLE
Fetch client dependant data server-side

### DISCUSSION THREAD

**Discussion by niklasp** [vercel/next.js#55099] [App Router]
I am writing an app where one very fundamental thing for data fetching can be changed in the client: The API endpoint, and wonder how its done best

Background: The app lets the user pick from a dropdown a network they want to be connected with. Depending on that choice, almost all data in client components changes. I would like to do data fetching for all pages / components server-side. However I dont know how to use app router to prefetch data that depend on that one client setting.

I have tried an app wide state storing the `endpoint`. So thats always available in the client.

**Comment by devjiwonchoi** [comment]
I'm assuming that [Fetching Data on the Client with Route Handlers](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers) is what you're looking for.

**Comment by niklasp** [comment]
For me the prefetching / server-side-rendering is the important part. When I hardcode the connected API endpoint, the needed data can be fetched in the server-side-component and sent to the client. When I however make the API endpoint dynamic I have to convert (all) my components to client-side-components?

Put differently. I need N versions of server components = 1 for each supported API endpoint my app uses. Is that possible? E.g. can I "store" the endpoint a user selected on the server and make the server-component render with that stored endpoint? Then change it via an api call from client -> server and then the server-rendered component updates?

The code I am using is this in a `server` component.

```ts
const getReferenda = async () => {
  const { api } = await getApi(endpoints[0]);
  const referenda = await api?.getReferenda();
  return referenda
};

export default async function ReferendumList() {
  const referenda = await getReferenda();

  return (
    <div className="referendum-list">
      {referenda?.map((ref) => {
        return (
          <ReferendumDetail referendum={ref} />
        );
      })}
    </div>
  );
}
```
The user can now change the API endpoint in the UI (and thereby a context value) and that should e.g. change the `referenda` E.g. `endpoints[0]` above should be dynamic, but as far as I can tell, the server can not at all access a dynamic value?`ReferendumDetail` is a client component

**Reply by devjiwonchoi** [reply]
Based on my understanding, you want `getReferenda` to be serverside, `referenda` result to be dynamic based on the user's choice from the client component, and `ReferendumList` to be a server component?

**Reply by niklasp** [reply]
Thanks for the answer. Yeah that is my understanding. The question however is how to make `referenda` dynamic. After reading [this](https://github.com/vercel/next.js/discussions/54075) and the docs again, I think I can go with cookies?!

E.g. when a user switches the `endpoint` they want to be connected to in the frontend, I set a cookie via a `Server Action`, then `getReferenda` reads that cookie that contains the `endpoint` the referenda should be fetched for.

Trying to rephrase my question:

**How can I have one route with multiple Server Side Rendered (aka data prefetched) components? All referenda for endpoint 1 prefetched server side + all referenda for endpoint N prefetched server side**

It would be really nice to have examples on use cases like that in some demos or tutorials.

**Reply by devjiwonchoi** [reply]
Have you considered using dynamic routes or searchParams? You can pass [`params` context to get dynamic routes](https://nextjs.org/docs/app/api-reference/file-conventions/route#parameters) or use [`NextRequest.nextUrl.searchParams`](https://nextjs.org/docs/app/api-reference/functions/next-request#nexturl), if I haven't misunderstood your needs.

**Reply by niklasp** [reply]
So currently the route that shows the `ReferendumList` is `/vote`, you mean to have a route like `/vote/1`, `/vote/2` and so on based on the api endpoint we are using? The problem is that all routes will use different endpoints and I do not necessarily want to clutter the url with the endpoint. Are there options?

**Reply by devjiwonchoi** [reply]
If so, I guess you won’t like the searchParams option too right? (e.g. ?a=b) How about using headers?

**Reply by niklasp** [reply]
No indeed not, but I dig deeper in the docs finding that

> Next.js provides helpful functions you may need when fetching data in Server Components such as [cookies](https://nextjs.org/docs/app/api-reference/functions/cookies) and [headers](https://nextjs.org/docs/app/api-reference/functions/headers). These will cause the route to be dynamically rendered as they rely on request time information.

=> So every time I use cookies or headers there will be no static rendering? Or can I try with react's `cache`

I also thought about fetching the data from all endpoints in a `Promise.all` server side and then filter it in the client components based on the user's chosen endpoint. But the problem here is that the connection to the apis (endpoints) is a very costly operation that takes about 5seconds and I do not want to do that for N endpoints server side.

And then the docs also offer `generateStaticParams` 

> The generateStaticParams function can be used in combination with [dynamic route segments](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) to [statically generate](https://nextjs.org/docs/app/building-your-application/rendering/server-components#static-rendering-default) routes at build time instead of on-demand at request time.

But there is so much going on, I think I need to create a new post and tag some officials for help?
